rpc/server.go

```diff
@@ -116,6 +110,64 @@ func marshalResponse(response any, log *zerolog.Logger) []byte {
 	return responseData
 }
 
+type validationResult struct {
+	Error  []byte
+	Method string
+	Id     uint64
+}
+
+func validateRequest(requestData []byte, logger *zerolog.Logger) validationResult {
+	var request map[string]interface{}
+	vr := validationResult{}
+	err := json.Unmarshal(requestData, &request)
+	if err != nil {
+		vr.Error = marshalResponse(unexpectedRequestUnmarshalError, logger)
+		return vr
+	}
+
+	// jsonrpc spec says id can be a string, number.
+	// We only support numbers.
+	// When golang unmarshals JSON into an interface value, float64 is used for numbers.
+	requestId := request["id"]
+	if !assertType[float64](request["id"]) {
+		vr.Error = marshalResponse(invalidRequestError, logger)
+		return vr
+	}
+
+	// is the id an integer?
+	fRequestId := requestId.(float64)
+	if fRequestId != float64(uint64(fRequestId)) {
+		vr.Error = marshalResponse(invalidRequestError, logger)
+		return vr
+	}
+	vr.Id = uint64(fRequestId)
+
+	if !assertType[string](request["jsonrpc"]) || request["jsonrpc"] != "2.0" {
+		requestError := invalidRequestError
+		requestError.Id = vr.Id
+		vr.Error = marshalResponse(requestError, logger)
+		return vr
+	}
+
+	if !assertType[string](request["method"]) {
+		requestError := invalidRequestError
+		requestError.Id = vr.Id
+		vr.Error = marshalResponse(requestError, logger)
+		return vr
+	}
+	vr.Method = request["method"].(string)
+
+	return vr
+}
+
+func assertType[T string | float64](i interface{}) bool {
+	if i == nil {
+		return false
+	}
```

Comment:
I'm not sure I understand this special case. What happens if you remove it?

Reply:
Good catch, I am overthinking this part of the code base. Removed in f18992f7.

Reply:
Did you actually remove it? Looks like you just removed the use of it, but left it there and added a unit test?

Reply:
Good catch! Fixed in f3b6b840.